adding documentation for hpa custom metrics (#4496)
knative/docs: docs/serving/autoscaling/autoscaling-metrics.md

For more information about KPA and HPA, see the documentation on [Supported Autoscaler types](autoscaler-types.md).

* **Per-revision annotation key:** `autoscaling.knative.dev/metric`
* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `"cpu"`, `"memory"`, and `"custom"` metrics are only supported on revisions that use the HPA class.
* **Default:** `"concurrency"`

> === Per-revision concurrency configuration

Reviewer: ```suggestion
=== "Per-revision concurrency configuration"
```

> ```yaml
> apiVersion: serving.knative.dev/v1
> kind: Service
> metadata:
>   name: helloworld-go
>   namespace: default
> spec:
>   template:
>     metadata:

Reviewer: ```suggestion
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
```

>       annotations:
>         autoscaling.knative.dev/metric: "concurrency"
> ```

Reviewer: ```suggestion
          annotations:
            autoscaling.knative.dev/metric: "concurrency"
```
```

> === Per-revision rps configuration

Reviewer: ```suggestion
=== "Per-revision rps configuration"
```

> ```yaml
> apiVersion: serving.knative.dev/v1
> kind: Service
> metadata:
>   name: helloworld-go
>   namespace: default
> spec:
>   template:
>     metadata:
>       annotations:
>         autoscaling.knative.dev/metric: "rps"
> ```

Reviewer: ```suggestion
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
          annotations:
            autoscaling.knative.dev/metric: "rps"
    ```
```

> === Per-revision cpu configuration

Reviewer: ```suggestion
=== "Per-revision cpu configuration"
```

> ```yaml
> apiVersion: serving.knative.dev/v1
> kind: Service
> metadata:
>   name: helloworld-go
>   namespace: default
> spec:
>   template:
>     metadata:
>       annotations:
>         autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
>         autoscaling.knative.dev/metric: "cpu"
> ```

Reviewer: ```suggestion
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
          annotations:
            autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
            autoscaling.knative.dev/metric: "cpu"
    ```
```

> === Per-revision memory configuration

Reviewer: ```suggestion
=== "Per-revision memory configuration"
```

For more information about KPA and HPA, see the documentation on [Supported Autoscaler types](autoscaler-types.md).

* **Per-revision annotation key:** `autoscaling.knative.dev/metric`

> * **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `cpu`/`memory`/custom metrics are only supported on revisions with the HPA class.

Reviewer 2: ```suggestion
* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `cpu`/`memory`/`custom` metrics are only supported on revisions with the HPA class.
```

Author: :+1:

> * **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `cpu`/`memory`/`custom` metrics are only supported on revisions with the HPA class.

Reviewer 3: ```suggestion
* **Possible values:** `"concurrency"`, `"rps"`, `"cpu"`, `"memory"` or any custom metric name, depending on your Autoscaler type. The `"cpu"`, `"memory"`, and `"custom"` metrics are only supported on revisions that use the HPA class.
```

Author: :+1:

* **Default:** `"concurrency"`

> === Per-revision concurrency configuration

Reviewer 3: @snneji can you confirm whether it was formatted like this before for tabs or something? I can't find it documented in the style guide.

Reviewer: Yes this used to be formatted as tabs. The 4 spaces of indentation these code snippets used to have needs to added back for the formatting to render correctly.

  namespace: default
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/metric: "concurrency"
```

=== Per-revision rps configuration

```yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
  namespace: default
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/metric: "rps"
```

=== Per-revision cpu configuration

```yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
  namespace: default
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
        autoscaling.knative.dev/metric: "cpu"
```

=== Per-revision memory configuration

```yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
  namespace: default
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
        autoscaling.knative.dev/metric: "memory"
```

=== Per-revision custom metric configuration
Assuming that you have a custom metric named `metricName` you can create an HPA to autoscale the revision by that metric.

> The HPA will be configured to use the average value of that metric over all the pods of the revision.

Reviewer 2: ```suggestion
The HPA will be configured to use the **average value** of that metric over all the pods of the revision.
```

Author: :+1:

```yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
  namespace: default
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
        autoscaling.knative.dev/metric: "metricName"

> ```

Author: :+1:

      template:
        metadata:
          annotations:
            autoscaling.knative.dev/class: "hpa.autoscaling.knative.dev"
            autoscaling.knative.dev/metric: "memory"
    ```

> === Per-revision custom metric configuration

Reviewer: ```suggestion
=== "Per-revision custom metric configuration"
```
